src/rpc/connectivity/cluster.cc

```diff
@@ -963,16 +981,28 @@ void connectivity_cluster_t::run_t::handle(
     server_id_t remote_server_id;
     {
         if (deserialize_universal_and_check(conn, &remote_server_id, peername)) {
-            return;
+            return join_result_t::TEMPORARY_ERROR;
+        }
+
+        if (expected_server_id && expected_server_id != remote_server_id) {
+            if (peer_addr.ip().is_loopback()) {
+                return join_result_t::TEMPORARY_ERROR;
+            }
+
+            logINF("Rejected a connection from server %s since it does not have expected id %s.",
```

Comment:
I feel like we should still send a `fail_handshake` with an appropriate `handshake_result_t::error` here.